### PR TITLE
feat(traces): add Span name, kind and timing accessors

### PR DESCRIPTION
**Pull request by jpkrohling**
Closes E-2945. Sub-issue 5 of E-2940 (v0.1.0 API realignment).

## Motivation

overstory's internal-spans detector needs span durations, which is the whole
reason it carries a wire parser: `readSpanFields` in
`internal/detector/wire.go` on its unmerged wire-path branch, reading `name`
(5), `kind` (6), `start_time_unix_nano` (7) and `end_time_unix_nano` (8).
`Span` exposed only the three identifiers.

## What this adds

```go
func (s Span) Name() ([]byte, error)
func (s Span) Kind() (int32, error)
func (s Span) StartTimeUnixNano() (uint64, error)
func (s Span) EndTimeUnixNano() (uint64, error)
```

Purely additive. No existing accessor changes behavior or cost, and `wire.go`
is byte-identical to `main`.

`name` is returned as raw bytes, consistent with `Metric.Name`, and is **not**
checked for UTF-8 validity — pdata accepts an invalid-UTF-8 span name, so
rejecting one here would make the wire path stricter than the path it mirrors.
`kind` is `int32` for the same reason `SeverityNumber` is: an unexpected
negative enum value stays distinguishable from the defined OTLP range.

## The design decision

All four are singular scalars that pdata resolves last-value-wins over a full
unmarshal, so they share one schema-aware walk, `parseSpanFields`, and can
never disagree with each other about whether a payload is valid.

**The identifier accessors deliberately do not join that walk**, and this was
the main open question on the issue. Converging them would collapse two parser
classes into one — the thing `operations.md` lists as the "Accessor divergence"
risk. But last-value-wins requires walking every span to its end, and the only
behavior that changes is on malformed input: a conformant producer emits each
singular field exactly once, so first-match and last-wins return the same value
*and* the same verdict. Paying a cost on 100% of conformant spans to change
what happens on corrupt ones is the wrong trade.

The cost is also not small, and it is producer-shaped in a way that a naive
benchmark hides. pdata's `MarshalProto` writes fields **back-to-front**, so a
collector-marshalled span carries `trace_id` last and a first-match scan
already walks nearly the whole span. Stock protobuf — protobuf-go, Java, Python,
i.e. the OTLP SDK exporters — emits ascending order with `trace_id` first.
`Span.TraceID()` on one ascending-order span, full walk versus first-match:

| attributes | first-match | full walk | ratio |
|---|---:|---:|---:|
| 5 | 13.0 ns | 133.8 ns | 10x |
| 50 | 10.7 ns | 619.0 ns | 58x |
| 200 | 11.6 ns | 1,948 ns | 167x |

The accepted consequence is that `Span` carries two parser classes: on
malformed bytes `Name()` can reject a span whose `TraceID()` succeeds.
`TestSpanIdentifiers_FirstMatchDivergence` pins it, `docs/DESIGN.md` records the
trade, and `operations.md` tabulates every wire-versus-pdata disagreement.

`parseSpanFields` is schema-aware at the Span level and framing-only below it:
`attributes`, `events`, `links` and `status` are wire-type- and
containment-checked but not parsed. `parseLogRecordSeverity` *does* descend into
the body and attributes; the discriminator is what the accessor's own consumer
reads next — severity consumers read a record's attributes, and sage's parity
requirement made validating them load-bearing, whereas none of the Span fields
here depends on a span's attributes, events or links.

## Evidence

**Gates** — all clean on this head:

| Gate | Result |
| --- | --- |
| `go build ./...` | pass |
| `go vet ./...` | pass |
| `go test -race -shuffle=on -count=1 ./...` | pass |
| `golangci-lint run` | 0 issues |
| `gofmt -l .` | clean |
| `git diff --check` | clean |
| `go mod tidy -diff` | fails identically on clean `upstream/main` (pre-existing `go.sum` checksum history); this branch touches neither `go.mod` nor `go.sum`. Per the exception in `CONTRIBUTING.md` and E-2940 |

**Tests.** `span_fields_test.go` adds pdata differential fixtures for absent,
zero-value, populated, descending-order, repeated-scalar, out-of-range kind,
negative-truncating kind, maximum timestamps, invalid-UTF-8 name, unknown-field
and all-known-fields-present shapes; malformed parity across the four scalar
accessors over the whole 16-field Span schema; identifier size validation;
trailing corruption; the absent-versus-empty name distinction pdata cannot
represent; the aliasing and capacity-clamp contract; and a
`testing.AllocsPerRun` gate on all seven accessors.

**Mutation testing drove real hardening.** Across two passes (the walk was
rewritten between them), every mutation is now killed. Four survived at some
point and each exposed a genuine gap:

- Removing the `flags` wire-type check survived, because every wrong-wire-type
  fixture happened to desynchronize the walk and error through framing instead
  — the checks themselves were untested. Added
  `TestSpanFields_WireTypeChecksAreLoadBearing`: 13 fixtures encoding a known
  field with the wrong wire type but a payload the *expected* reader consumes
  whole, so the walk stays aligned and a missing check yields a silently
  successful parse.
- Dropping the `parentSpanID` zero-reset survived because the trailing-empty
  fixture covered only fields 1 and 2.
- Dropping the `num == 5` guard, so `trace_state` also wrote `name`, survived
  all 47 fixtures: the only `trace_state` fixture emitted field 3 *before*
  field 5, so last-wins still landed on the name. Added a `trace_state after
  name` fixture.
- A walk that stopped at `flags` (16) survived, because no fixture placed any
  field after it. That matters precisely because pdata's reverse order puts the
  three identifiers *after* flags. Added a `fields after flags` fixture.

**Divergences were verified, not assumed.** Every claim in the `operations.md`
table was probed against pdata before being written down: malformed contents
inside `attributes`/`events`/`links`/`status` are accepted here and rejected by
pdata; an unknown group closing at the very end of the span is accepted here
and rejected by pdata; and — running the other way — an unknown group closed
with a field number that does not match its StartGroup is rejected here and
accepted by pdata, because `protowire` compares the numbers and pdata's
`ConsumeUnknown` only counts depth. The previous wording of the group row was
imprecise: a matched group mid-span is accepted by both.

**Benchmarks.** `docs/BENCHMARKS.md` has the full section with commands and
environment. Paired comparisons are alternating invocations of prebuilt
binaries, reported as the median of paired per-round deltas plus the sign
count, per the `operations.md` release checklist.

Identifier accessors, against a merge-base worktree checkout of `7a777b5`:
**−4.3% on `BenchmarkSpanIteration`, slower in 6 of 21 rounds** — noise around
zero, which is the expected result for byte-identical code.

The four scalar accessors, 5 resources × 100 spans with five attributes each:

| Arm | ns/op | B/op | allocs/op |
|---|---:|---:|---:|
| `HandRolled` (overstory's branch) | 86,236 | 552 | 24 |
| `Accessors` | 213,335 | 552 | 24 |
| `Unmarshal` (pdata) | 698,032 | 558,816 | 10,456 |
| `TraceIDOnly` (first-match, control) | 68,486 | 528 | 24 |

**+151%** versus the hand-rolled walk, slower in 15 of 15, at allocation
parity. The gap is the walk count: four scalar accessor calls, four walks,
against one.

Read that against the right baseline. overstory runs **pdata unmarshal** on
`main` today; the hand-rolled walk is on an unmerged branch. Migrating from
pdata to these accessors is a **3.3x improvement** and drops 10,456 allocations
per batch to 24. Migrating from the hand-rolled walk is a 2.5x cost. Which
applies depends on which branch overstory migrates from — that belongs on the
migration issue.

## Risks and rollout

- **Risk: low.** Purely additive; no existing code path changed, and `wire.go`
  is untouched. The only shared-code edits are documentation.
- No consumer action required. overstory's migration is tracked separately per
  the issue's acceptance criteria.
- Once overstory migrates, drop `readSpanFieldsHandRolled` from
  `benchmark_comparison_test.go` and its `docs/BENCHMARKS.md` subsection — it
  will then benchmark code that exists nowhere. Noted in both places.
- **No combined accessor is proposed here, but the numbers argue for one.**
  `parseSpanFields` already produces all four scalars per walk, so exposing
  them together needs no new parsing machinery — only a decision on the
  exported shape, which depends on which fields overstory actually takes.
  Deferred to its migration, not declined.
- No tag or release proposed. Per E-2940, the next version should wait for the
  rest of the sub-issue set (E-2946 still open).

## Agent involvement

Per `CONTRIBUTING.md`: material. Implementation, tests, benchmarks and
documentation were agent-generated. Four sub-agents reviewed the diff for
reuse, simplification, efficiency and altitude; a fifth ran an adversarial pass
that found the producer-order problem above, two of the four test blind spots,
and the group-mismatch divergence. Two mutation passes over `parseSpanFields`
drove the test hardening. A human contributor reviews and owns this before
merge.

## Summary by CodeRabbit

- **New Features**
  - Added Span accessors for name, kind, start time, and end time.
  - Added schema-aware validation for span fields, including malformed data and invalid wire types.
  - Added an example that calculates and reports durations for internal spans.

- **Documentation**
  - Documented Span accessors, validation behavior, performance characteristics, and malformed-input handling.
  - Added benchmark documentation comparing accessor performance and allocations.

- **Tests**
  - Expanded coverage for absent fields, repeated values, invalid data, parity, edge cases, and allocation behavior.

### CODE REVIEW

File: docs/BENCHMARKS.md [hunk 1]
```diff
@@ -680,3 +680,100 @@ fields per record.
...
+  done
+done
+```
+
+| Benchmark | ns/op (median) | B/op | allocs/op |
+|---|---:|---:|---:|
+| `BenchmarkSpan_InternalSpans_HandRolled` | 86,236 | 552 | 24 |
+| `BenchmarkSpan_InternalSpans_Accessors` | 213,335 | 552 | 24 |
+| `BenchmarkSpan_InternalSpans_Unmarshal` (pdata) | 698,032 | 558,816 | 10,456 |
+| `BenchmarkSpan_TraceIDOnly` (first-match, control) | 68,486 | 528 | 24 |
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply docs/BENCHMARKS.md:756]: Fixed in 7579b94. The command block now covers both comparisons separately: 15 alternating rounds for the `_HandRolled` vs `_Accessors` paired delta, and a 5-round rotation over all four benchmarks for the reference medians (`_Unmarshal` and `BenchmarkSpan_TraceIDOnly`), which are reference points rather than arms of the paired comparison. The table is now reproducible row by row.

File: docs/BENCHMARKS.md [hunk 2]
```diff
@@ -680,3 +680,100 @@ fields per record.
...
+
+**About 151% slower than the hand-rolled walk** (median paired delta +151.4%),
+slower in 15 of 15 rounds, at **allocation parity**. The gap is the walk count:
+four scalar accessor calls, four walks, against one.
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply docs/BENCHMARKS.md]: Correct, and this was a real error in the text. The reduction calls `TraceID()` and `Kind()` on every span, then `Name()` and the two timestamps only after the internal-span filter passes. The fixture cycles kinds through five values of which two are `Internal`, so the arm averages 1 + 0.4 x 3 = 2.2 scalar walks per span, not four. Fixed in 7579b94, which also notes that a consumer reading all four scalars unconditionally would pay four.

File: docs/DESIGN.md [hunk 3]
```diff
@@ -167,10 +168,49 @@ E-2941) previously played that role directly; its replacement is
...
+the divergence risk in [operations.md](../operations.md), taken knowingly here
+rather than engineered away, and pinned by
+`TestSpanIdentifiers_FirstMatchDivergence` so it cannot drift unnoticed.
+
+**The walk is framing-only below the Span level.** It checks the wire type and
+containment of `attributes`, `events`, `links` and `status` without parsing
+their contents, where the LogRecord walk does descend into the body and every
+attribute. The discriminator is what the accessor's own consumer reads next: a
+severity consumer reads the record's attributes, and sage's parity requirement
+made validating them load-bearing, whereas none of the Span fields this package
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply docs/DESIGN.md:204]: Skipping this one. Consumer names are written lowercase throughout the prose in `docs/DESIGN.md` and `docs/BENCHMARKS.md` — `sage`, `mulch`, `marigold`, and `overstory` in this change — a convention established by #28 and #29. The capitalised forms appear only in the consumer tables in `docs/specification.md`. Capitalising this one instance would make it the outlier. (The proposed diff is also a no-op: the before and after lines are identical.)

File: operations.md [hunk 4]
```diff
@@ -22,7 +22,7 @@ this before releasing the module or changing a consumer rollout boundary.
...
-| Accessor divergence on one message | hypothesized | Two accessors over the same message disagree on whether a payload is valid, so a consumer sees one field read succeed and the other fail on identical bytes | A second accessor added with its own parser instead of reading from the existing walk | Shared-walk implementation plus a malformed-parity test asserting both accessors fail together; see `TestLogRecordSeverity_MalformedParity` |
+| Accessor divergence on one message | hypothesized | Two accessors over the same message disagree on whether a payload is valid, so a consumer sees one field read succeed and the other fail on identical bytes | A second accessor added with its own parser instead of reading from the existing walk, or a new accessor scanning a message whose existing accessors scan differently | Shared-walk implementation plus a malformed-parity test asserting every accessor over that message fails together; see `TestLogRecordSeverity_MalformedParity` and `TestSpanFields_MalformedParity` |
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply operations.md]: Both parts valid and fixed in 7579b94.

The `operations.md` row now says the malformed-parity tests assert that *accessors sharing a schema-aware walk* fail together, and names `TestSpanIdentifiers_FirstMatchDivergence` as the pin for the case where accessors over one message deliberately do not share a walk.

The `docs/specification.md` sentence was worse than ambiguous — it was left over from a shared-walk design that was reverted before this PR was opened, and described behaviour the shipped code does not have. It now states that the identifier accessors select the first matching occurrence and stop, so an empty occurrence yields zero only when it is that first one, and later occurrences are never examined. Good catch.

**Review by jpkrohling — COMMENTED** [review]
human review ok, lgtm
